### ISSUE TITLE
cuDNN 5.1 Compilation Error

### ISSUE THREAD

**Issue by jessejohns** [Theano/Theano#5009]
Hello,

I am trying to use the models here:

https://github.com/titu1994/Inception-v4

With theano '0.9.0dev2.dev-754f3b721e306b80a0d8554d93ae2d43076b4735' using CUDA 8, cuDNN 5.1 on a 1080 GTX, I receive the following error:

```
RuntimeError: ('The following error happened while compiling the node', GpuDnnConvGradW{algo='none', inplace=True}(GpuContiguous.0, GpuContiguous.0, GpuAllocEmpty{dtype='float32', context_name=None}.0, GpuDnnConvDesc{border_mode='half', subsample=(1, 1), conv_mode='conv', precision='float32'}.0, Constant{1.0}, Constant{0.0}), '\n', 'could not create cuDNN handle: CUDNN_STATUS_INTERNAL_ERROR')
```

However on a K80 with theano '0.8.2' using CUDA 7.5 and cuDNN 4, I can compile and run the model just fine.

**Comment by nouiz** [comment]
Did you update the driver? You need a recent driver for that GPU.  If you
don't use cudnn, does it work?

**Comment by jessejohns** [comment]
Yes, I have the latest version of the driver: 367.35   I have no issues compiling any other models with the 1080.  

With cuDNN, I get the following errors. I'll take a shot at changing up the model to get rid of the apparent dependencies.

```
ERROR (theano.gof.opt): Optimization failure due to: local_gpua_pool_dnn_alternative
ERROR (theano.gof.opt): node: Pool{ignore_border=True, mode='average_exc_pad'}(Elemwise{switch,no_inplace}.0, TensorConstant{(2,) of 3}, TensorConstant{(2,) of 1}, TensorConstant{(2,) of 1})                                    
ERROR (theano.gof.opt): TRACEBACK:                                                                               
ERROR (theano.gof.opt): Traceback (most recent call last):                                                       
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gof/opt.py", line 1922, in process_node            
    replacements = lopt.transform(node)                                                                          
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gpuarray/opt.py", line 204, in local_opt           
    new_op = maker(node.op, context_name, node.inputs, node.outputs)                                             
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gpuarray/dnn.py", line 1867, in local_gpua_pool_dnn_alternative                                                                                                     
    raise_no_cudnn()                                                                                             
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gpuarray/dnn.py", line 44, in raise_no_cudnn       
    raise RuntimeError(msg)                                                                                      
RuntimeError: cuDNN is required for convolution and pooling  
```

```
Traceback (most recent call last):                                                                               
  File "main.py", line 193, in <module>                                                                          
    model = mk.cnn(network, params, weights_load=weights_load)                                                   
  File "/home/jesse/projects/hv/flag/scripts/model_keras.py", line 52, in cnn                                    
    model = inception_v4(params,weights_load=weights_load)                                                       
  File "/home/jesse/projects/hv/flag/scripts/model_keras.py", line 1141, in inception_v4                         
    output = create_model(inputs,n_class)                                                                        
  File "/home/jesse/projects/hv/flag/scripts/model_keras.py", line 1097, in create_model                         
    x = inception_stem(inputs)
  File "/home/jesse/projects/hv/flag/scripts/model_keras.py", line 1002, in inception_stem
    m3 = BatchNormalization(axis=1)(m3)
  File "/home/jesse/.local/lib/python2.7/site-packages/keras/engine/topology.py", line 515, in __call__
    self.add_inbound_node(inbound_layers, node_indices, tensor_indices)
  File "/home/jesse/.local/lib/python2.7/site-packages/keras/engine/topology.py", line 573, in add_inbound_node
    Node.create_node(self, inbound_layers, node_indices, tensor_indices)
  File "/home/jesse/.local/lib/python2.7/site-packages/keras/engine/topology.py", line 150, in create_node
    output_tensors = to_list(outbound_layer.call(input_tensors[0], mask=input_masks[0]))
  File "/home/jesse/.local/lib/python2.7/site-packages/keras/layers/normalization.py", line 122, in call
    epsilon=self.epsilon)
  File "/home/jesse/.local/lib/python2.7/site-packages/keras/backend/theano_backend.py", line 382, in normalize_batch_in_training
    epsilon)
  File "/home/jesse/.local/lib/python2.7/site-packages/keras/backend/theano_backend.py", line 392, in batch_normalization
    'spatial', epsilon)
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/sandbox/cuda/dnn.py", line 2793, in dnn_batch_normalization_test
    batchnorm_op = GpuDnnBatchNormInference(mode=mode, epsilon=epsilon)
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/sandbox/cuda/dnn.py", line 2312, in __init__
    if version() < (5000, 5000):
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/sandbox/cuda/__init__.py", line 417, in dnn_version
    dnn_available.msg)
Exception: ("We can't determine the cudnn version as it is not available", 'Disabled by dnn.enabled flag')
```

**Comment by nouiz** [comment]
Update Theano to the dev version. There is many changes since the last
release.

**Comment by jessejohns** [comment]
With version: '0.9.0dev2.dev-5cbbcb72e254e1bb26a479aeecbf64f4ce3f6eaa'

The follow error still persists:

```
File "/home/jesse/.local/lib/python2.7/site-packages/theano/compile/function.py", line 326, in function                                                         
    output_keys=output_keys)                                                     
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/compile/pfunc.py", line 484, in pfunc                                                               
    output_keys=output_keys)                                                     
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/compile/function_module.py", line 1789, in orig_function                                            
    defaults)                                                                    
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/compile/function_module.py", line 1653, in create                                                   
    input_storage=input_storage_lists, storage_map=storage_map)                  
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gof/link.py", line 699, in make_thunk
    storage_map=storage_map)[:3]
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gof/vm.py", line 1051, in make_all
    no_recycling))
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gof/op.py", line 932, in make_thunk
    no_recycling)
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gof/op.py", line 833, in make_c_thunk
    output_storage=node_output_storage)
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gof/cc.py", line 1190, in make_thunk
    keep_lock=keep_lock)
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gof/cc.py", line 1131, in __compile__
    keep_lock=keep_lock)
  File "/home/jesse/.local/lib/python2.7/site-packages/theano/gof/cc.py", line 1606, in cthunk_factory
    *(in_storage + out_storage + orphd))
RuntimeError: ('The following error happened while compiling the node', GpuDnnConvGradI{algo='none', inplace=True}(GpuContiguous.0, GpuContiguous.0, GpuAllocEmpty{dtype='float32', context_name=None}.0, GpuDnnConvDesc{border_mode='half', subsample=(1, 1), conv_mode='conv', precision='float32'}.0, Constant{1.0}, Constant{0.0}), '\n', 'could not create cuDNN handle: CUDNN_STATUS_INTERNAL_ERROR')
```

**Comment by jessejohns** [comment]
With the latest CUDA/cudnn and theano, I get an error that is more informative:
```

Traceback (most recent call last):                                               
  File "main.py", line 218, in <module>                                          
    model = mk.cnn(network, params, weights_load=weights_load)                   
  File "/home/jesse/projects/hv/scripts/model_keras.py", line 50, in cnn         
    model = inception_resnet_v2(params,weights_load=weights_load)                
  File "/home/jesse/projects/hv/scripts/model_keras.py", line 1207, in inception_resnet_v2                                                                        
    output = create_model(inputs,n_class)                                        
  File "/home/jesse/projects/hv/scripts/model_keras.py", line 1157, in create_model                                                                               
    x = inception_resnet_stem(inputs)                                            
  File "/home/jesse/projects/hv/scripts/model_keras.py", line 1058, in inception_resnet_stem                                                                      
    m3 = BatchNormalization(axis=1)(m3)                                          
  File "/home/jesse/Programs/anaconda/lib/python2.7/site-packages/keras/engine/topology.py", line 514, in __call__                                                
    self.add_inbound_node(inbound_layers, node_indices, tensor_indices)          
  File "/home/jesse/Programs/anaconda/lib/python2.7/site-packages/keras/engine/topology.py", line 572, in add_inbound_node                                        
    Node.create_node(self, inbound_layers, node_indices, tensor_indices)         
  File "/home/jesse/Programs/anaconda/lib/python2.7/site-packages/keras/engine/topology.py", line 149, in create_node                                             
    output_tensors = to_list(outbound_layer.call(input_tensors[0], mask=input_masks[0]))                                                                          
  File "/home/jesse/Programs/anaconda/lib/python2.7/site-packages/keras/layers/normalization.py", line 138, in call                                               
    epsilon=self.epsilon)                                                        
  File "/home/jesse/Programs/anaconda/lib/python2.7/site-packages/keras/backend/theano_backend.py", line 415, in normalize_batch_in_training                      
    epsilon)                                                                     
  File "/home/jesse/Programs/anaconda/lib/python2.7/site-packages/keras/backend/theano_backend.py", line 428, in batch_normalization
    'spatial', epsilon)
  File "/home/jesse/Programs/anaconda/lib/python2.7/site-packages/theano/sandbox/cuda/dnn.py", line 2793, in dnn_batch_normalization_test
    batchnorm_op = GpuDnnBatchNormInference(mode=mode, epsilon=epsilon)
  File "/home/jesse/Programs/anaconda/lib/python2.7/site-packages/theano/sandbox/cuda/dnn.py", line 2312, in __init__
    if version() < (5000, 5000):
  File "/home/jesse/Programs/anaconda/lib/python2.7/site-packages/theano/sandbox/cuda/__init__.py", line 417, in dnn_version
    dnn_available.msg)
Exception: ("We can't determine the cudnn version as it is not available", 'CUDA not available')
```

So, removing the batch normalization layer, I can run the model.   Just a note, CUDA is indeed available.

**Comment by jihunchoi** [comment]
I am experiencing the similar issue on GTX 1070, cudnn 5.1, cuda 8.0.
It seems that errors occur when I call cudnn-related functions/methods.

For me, the error message is given as follows when using conv2d.

('The following error happened while compiling the node', GpuDnnConv{algo='small', inplace=True}(GpuContiguous.0, GpuContiguous.0, GpuAllocEmpty{
dtype='float32', context_name=None}.0, GpuDnnConvDesc{border_mode='valid', subsample=(1, 1), conv_mode='conv', precision='float32'}.0, Constant{1.0}, Const
ant{0.0}), '\n', 'cudnn_handle')

And the below is the error message when using newly merged cudnn RNN.
('The following error happened while compiling the node', _DropoutDescriptor{context_name=None}(Cast{float32}.0, Cast{uint64}.0, Constant{<pygpu.gpu...b39314930>}), '\n', 'cudnn_handle')


I also tried to use cudnn 5.0 instead of 5.1 but the same error messages occurred...

**Comment by carlthome** [comment]
Similar issue here (latest Theano build from master, latest cudnn and display drivers AFAIK).

```sh
THEANO_FLAGS=dnn.enabled=True python -c "import theano"

Using gpu device 0: TITAN X (Pascal) (CNMeM is enabled with initial size: 95.0% of memory, cuDNN Can not compile with cuDNN. We got this error:
b"nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).\n/usr/bin/ld: /tmp/tmpxft_000078ff_00000000-4_try_flags_8lc4ag0a.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC\n/usr/bin/ld: final link failed: Nonrepresentable section on output\ncollect2: error: ld returned 1 exit status\n")
```

**Comment by angelsrates** [comment]
Hi! Did someone could fix it? I have the same error with Theano 0.9dev, CUDA 7.5 and cuDNN 5.1

**Comment by nouiz** [comment]
What is your GPU?

Make sure to use a good nvidia driver that support your GPU. The previous
error was using the new Titan X that need a very recent GPU.

**Comment by angelsrates** [comment]
My GPU is also the new Titan X (Pascal).
And drivers 365, cuDNN 5.1 (CUDA 7.5). Is it possible to install CUDA 8 for Ubuntu 14.04?

Thanks!

**Comment by nouiz** [comment]
You need cuda 8 for that new GPU. Check nvidia installer. Maybe the
released it too for ubuntu 14.04. Otherwise, people will probably write
blog post with work around on the web.

**Comment by nouiz** [comment]
as a solution was provided and no news, I'll close.
